### ISSUE TITLE
Classification Model for General Quantum Circuits

### ISSUE THREAD

**Issue by ShreyK7** [dorahacksglobal/qc-classifier#2]
We are looking to expand/repurpose our classifier model to classify output data from generalized quantum circuits in the same way we did for QRNG circuits (i.e. not just quantum random number generation circuits, but any kind of x qubit quantum circuit and/or algorithm that can be run through a quantum device to generate data). Feel free to use any ML/AI models/techniques as well as quantum theory or data processing/extraction functions, including those we have already included in this repo for classifying specifically QNRG data. We have not yet attempted the task for general circuits, so there is no benchmark classification accuracy as of now - in other words, any positive results are valuable! Additionally, if needed, feel free to generate quantum training/test data from any sources possible.

**Comment by sid-chava** [comment]
Just merged - looks good. Keep in mind that that the python random module is psuedo random as opposed to the quantum random numbers generated in the training file. Do you guys have access to the original training circuit somewhere? if not I can upload it to this repository so let me know.

**Comment by sid-chava** [comment]
> Hi! @0mWh and I would like to work on this too. What kind of general circuits are you looking for? Will benchmark circuits suffice, or should a more practical circuit like QAOA or quantum neural network be deployed?

Since we are effectively classifying the noise specific to select quantum computers, we are trying to see if we would be able to extend this to the output of other quantum circuits. So if you take other quantum circuits and run it over different quantum computers, would the model be able to separate the select quantum computers?

**Comment by 0mWh** [comment]
> Do you guys have access to the original training circuit somewhere? if not I can upload it to this repository so let me know.

no, I do not see any quantum circuits in the repository (are they somewhere else?)

**Comment by sid-chava** [comment]
> > Hi! @0mWh and I would like to work on this too. What kind of general circuits are you looking for? Will benchmark circuits suffice, or should a more practical circuit like QAOA or quantum neural network be deployed?
> 
> Since we are effectively classifying the noise specific to select quantum computers, we are trying to see if we would be able to extend this to the output of other quantum circuits. So if you take other quantum circuits and run it over different quantum computers, would the model be able to separate the select quantum computers?

@bernardyeszero should be able to provide more background on this when he gets online

**Comment by sid-chava** [comment]
> > Do you guys have access to the original training circuit somewhere? if not I can upload it to this repository so let me know.
> 
> no, I do not see any quantum circuits in the repository (are they somewhere else?)

https://github.com/dorahacksglobal/quantum-randomness-generator/blob/main/Rigetti_CHSH_pub.ipynb

I believe this repository should have most of the details that will be helpful for this part of the task. The Readme of this repository outlines the process of generating random numbers using the CHSH game in more detail.

**Comment by ShreyK7** [comment]
> Hi! @0mWh and I would like to work on this too. What kind of general circuits are you looking for? Will benchmark circuits suffice, or should a more practical circuit like QAOA or quantum neural network be deployed?

Just to add on - for now, the specific circuit itself doesn't matter, more so seeing if we can develop a classification model that can distinguish QC's (through circuit output data, most likely) through their device specific quantum noise. As @sid-chava mentioned, we are looking to see if we can classify any type of quantum circuit (beyond our preexisting QNRG generation circuit) with a prediction model. I believe @bernardyeszero will update the issue description sometime soon to add a little more detail (and I think he will specify some kind of 10 qubit general circuit), but for now, if it helps, feel free to use whatever quantum circuits you would like - the focus is classifying them and predicting which quantum computers they were run on.

**Comment by ShreyK7** [comment]
> Hi @ShreyK7 made a PR #4 To address this issue it covers Generate Synthetic Data: The generate_quantum_data() function creates synthetic data mimicking the output of generalized quantum circuits. In this example, it generates random numerical data for the features (X) and binary labels (y).
> 
> Define the Classifier Model: The build_classifier() function defines the classifier model. In this example, it uses a Random Forest classifier with 100 trees.
> 
> Data Preparation: Set the number of samples (num_samples) and features (num_features). Then, generate synthetic quantum data using the generate_quantum_data() function.
> 
> Split Data: Use train_test_split() from scikit-learn to split the data into training and testing sets. Here, 80% of the data is used for training (X_train, y_train), and 20% for testing (X_test, y_test).
> 
> Build and Train Model: Create an instance of the classifier model using build_classifier() and train it on the training data (X_train, y_train) using the fit() method.
> 
> Prediction: Use the trained classifier to predict labels for the testing set (X_test) using the predict() method.
> 
> Evaluate Model: Calculate the accuracy of the model's predictions using accuracy_score() from scikit-learn.
> 
> accuracy - Accuracy: 0.49

Yeah to add on to @sid-chava's point: we've already explored random forests in our QNRG classification notebook - your model looks well defined so can definitely build off of your initial methods but at the end of the day the important part of this task/issue/bounty is seeing what accuracy we can get when applying a ML/AI classification model to an actual general quantum circuit run through different QCs to generate output data.

**Comment by sid-chava** [comment]
> > Hi @ShreyK7 made a PR #4 To address this issue it covers Generate Synthetic Data: The generate_quantum_data() function creates synthetic data mimicking the output of generalized quantum circuits. In this example, it generates random numerical data for the features (X) and binary labels (y).
> > Define the Classifier Model: The build_classifier() function defines the classifier model. In this example, it uses a Random Forest classifier with 100 trees.
> > Data Preparation: Set the number of samples (num_samples) and features (num_features). Then, generate synthetic quantum data using the generate_quantum_data() function.
> > Split Data: Use train_test_split() from scikit-learn to split the data into training and testing sets. Here, 80% of the data is used for training (X_train, y_train), and 20% for testing (X_test, y_test).
> > Build and Train Model: Create an instance of the classifier model using build_classifier() and train it on the training data (X_train, y_train) using the fit() method.
> > Prediction: Use the trained classifier to predict labels for the testing set (X_test) using the predict() method.
> > Evaluate Model: Calculate the accuracy of the model's predictions using accuracy_score() from scikit-learn.
> > accuracy - Accuracy: 0.49
> 
> Yeah to add on to @sid-chava's point: we've already explored random forests in our QNRG classification notebook - your model looks well defined so can definitely build off of your initial methods but at the end of the day the important part of this task/issue/bounty is seeing what accuracy we can get when applying a ML/AI classification model to an actual general quantum circuit run through different QCs to generate output data.

Apologies for not catching this earlier, but since we have this functionality described already in the original notebook, we will roll this back to avoid general confusion.

**Comment by ShreyK7** [comment]
@AbdullahKazi500 submit a PR with your updated code and @bernardyeszero can then take a look at it

**Comment by ShreyK7** [comment]
Thanks hopefully @bernardyeszero can take a closer look at your work soon. One thing I will say is that you theoretically should not be getting 100% accuracy each team (this implies that your circuit backend to generate training data is doing something wrong).
